File: python-sdk/runtime_observer/exporter.py
from __future__ import annotations

import atexit
import json
import queue
import threading
import time
import urllib.error
import urllib.request
from typing import Any

from .config import RuntimeObserverConfig


class BatchExporter:
    def __init__(self, config: RuntimeObserverConfig):
        self.config = config
        self.queue: queue.Queue[dict[str, Any]] = queue.Queue(maxsize=config.max_queue_size)
        self._queue = self.queue
        self.dropped_events = 0
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None

    def start(self) -> None:
        if self._thread or not self.config.exporting_enabled:
            return
        self._thread = threading.Thread(target=self._run, name="runtime-observer-exporter", daemon=True)
        self._thread.start()
        atexit.register(self.shutdown)

    def enqueue(self, event: dict[str, Any]) -> None:
        if not self.config.exporting_enabled:
            return
        try:
            raw = json.dumps(event, default=str)
            if len(raw.encode("utf-8")) > self.config.max_event_size_bytes:
                event = {**event, "payload": {"truncated": True, "original_kind": event.get("kind")}}
            self.queue.put_nowait(event)
        except queue.Full:
            self.dropped_events += 1
            return

    def _run(self) -> None:
        while not self._stop.is_set():
            self.flush(timeout=self.config.flush_interval_seconds)

    def flush(self, timeout: float | None = None) -> None:
        batch: list[dict[str, Any]] = []
        deadline = time.time() + (timeout or 0)
        while len(batch) < self.config.batch_size:
            try:
                if not batch and timeout:
                    item = self.queue.get(timeout=max(0.0, deadline - time.time()))
                else:
                    item = self.queue.get_nowait()
                batch.append(item)
            except queue.Empty:
                break
        if batch:
            self._send(batch)

    def _send(self, events: list[dict[str, Any]]) -> None:
        url = self.config.endpoint.rstrip("/") + "/v1/ingest"
        data = json.dumps({"batch_id": events[0].get("event_id"), "events": events}, default=str).encode("utf-8")
        request = urllib.request.Request(
            url,
            data=data,
            headers={"Content-Type": "application/json", "Authorization": f"Bearer {self.config.api_key or 'local-dev-key'}"},
            method="POST",
        )
        try:
            with urllib.request.urlopen(request, timeout=2) as response:
                response.read()
        except (urllib.error.URLError, TimeoutError, OSError):
            return

    def shutdown(self, timeout: float = 5.0) -> None:
        self._stop.set()
        end = time.time() + timeout
        while not self.queue.empty() and time.time() < end:
            self.flush(timeout=0.05)
        if self._thread and self._thread.is_alive():
            self._thread.join(timeout=0.2)


Exporter = BatchExporter

File: js-sdk/core/exporter.js
import { exportingEnabled } from './index.js';

export class BatchExporter {
  constructor(config, { browser = false } = {}) {
    this.config = config;
    this.browser = browser;
    this.queue = [];
    this.droppedEvents = 0;
    this.timer = null;
  }

  start() {
    if (this.timer || !exportingEnabled(this.config)) return;
    this.timer = setInterval(() => {
      this.flush().catch(() => undefined);
    }, this.config.flushIntervalMs || 2000);
    if (typeof this.timer.unref === 'function') this.timer.unref();
  }

  enqueue(event) {
    if (!exportingEnabled(this.config)) return;
    let candidate = event;
    try {
      if (new TextEncoder().encode(JSON.stringify(candidate)).length > this.config.maxEventSizeBytes) {
        candidate = { ...event, payload: { truncated: true, original_kind: event.kind } };
      }
    } catch {
      candidate = { ...event, payload: { truncated: true, original_kind: event.kind } };
    }
    if (this.queue.length >= this.config.maxQueueSize) {
      this.droppedEvents += 1;
      return;
    }
    this.queue.push(candidate);
  }

  async flush() {
    if (!exportingEnabled(this.config) || this.queue.length === 0) return;
    const events = this.queue.splice(0, this.config.batchSize || 100);
    try {
      await this.send(events);
    } catch {
      // Match the Python SDK's best-effort behavior. Keep bounded memory.
    }
  }

  async send(events) {
    const endpoint = String(this.config.endpoint || '').replace(/\/$/, '');
    const body = JSON.stringify({ batch_id: events[0]?.event_id, events });
    if (this.browser) {
      const url = `${endpoint}/v1/ingest/browser?api_key=${encodeURIComponent(this.config.apiKey || '')}`;
      // Do not use sendBeacon: it always attaches cookies for the target origin, which turns
      // ingest into a credentialed cross-origin request and trips Access-Control-Allow-Origin: *.
      // fetch with keepalive covers the page-unload case and lets us opt out of credentials.
      const response = await fetch(url, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body,
        keepalive: true,
        credentials: 'omit',
        mode: 'cors',
      });
      if (!response.ok) throw new Error(`Runtime Observer ingest failed: ${response.status}`);
      return;
    }
    const response = await fetch(`${endpoint}/v1/ingest`, {
      method: 'POST',
      headers: { 'content-type': 'application/json', authorization: `Bearer ${this.config.apiKey || 'local-dev-key'}` },
      body,
    });
    if (!response.ok) throw new Error(`Runtime Observer ingest failed: ${response.status}`);
  }

  async shutdown() {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
    while (this.queue.length) await this.flush();
  }
}
